akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowIteratorSpec.scala

```diff
@@ -13,17 +12,139 @@ import akka.stream.testkit.StreamTestKit
 import akka.stream.testkit.StreamTestKit.OnComplete
 import akka.stream.testkit.StreamTestKit.OnError
 import akka.stream.testkit.StreamTestKit.OnNext
+import akka.stream.impl.SynchronousIterablePublisher
+import org.reactivestreams.Subscription
+import akka.testkit.TestProbe
+import org.reactivestreams.Subscriber
 
 class FlowIteratorSpec extends AbstractFlowIteratorSpec {
   override def testName = "A Flow based on an iterator producing function"
-  override def createSource[T](iterable: immutable.Iterable[T]): Source[T, Unit] =
-    Source(() ⇒ iterable.iterator)
+  override def createSource(elements: Int): Source[Int, Unit] =
+    Source(() ⇒ (1 to elements).iterator)
 }
 
 class FlowIterableSpec extends AbstractFlowIteratorSpec {
   override def testName = "A Flow based on an iterable"
-  override def createSource[T](iterable: immutable.Iterable[T]): Source[T, Unit] =
-    Source(iterable)
+  override def createSource(elements: Int): Source[Int, Unit] =
+    Source(1 to elements)
+
+  implicit def mmaterializer = super.materializer
+
+  "produce onError when iterator throws" in {
+    val iterable = new immutable.Iterable[Int] {
+      override def iterator: Iterator[Int] =
+        (1 to 3).iterator.map(x ⇒ if (x == 2) throw new IllegalStateException("not two") else x)
+    }
+    val p = Source(iterable).runWith(Sink.publisher())
+    val c = StreamTestKit.SubscriberProbe[Int]()
+    p.subscribe(c)
+    val sub = c.expectSubscription()
+    sub.request(1)
+    c.expectNext(1)
+    c.expectNoMsg(100.millis)
+    sub.request(2)
+    c.expectError.getMessage should be("not two")
+    sub.request(2)
+    c.expectNoMsg(100.millis)
+  }
+
+  "produce onError when Source construction throws" in {
+    val iterable = new immutable.Iterable[Int] {
+      override def iterator: Iterator[Int] = throw new IllegalStateException("no good iterator")
+    }
+    val p = Source(iterable).runWith(Sink.publisher())
+    val c = StreamTestKit.SubscriberProbe[Int]()
+    p.subscribe(c)
+    c.expectSubscriptionAndError().getMessage should be("no good iterator")
+    c.expectNoMsg(100.millis)
+  }
+
+  "produce onError when hasNext throws" in {
+    val iterable = new immutable.Iterable[Int] {
+      override def iterator: Iterator[Int] = new Iterator[Int] {
+        override def hasNext: Boolean = throw new IllegalStateException("no next")
+        override def next(): Int = -1
+      }
+    }
+    val p = Source(iterable).runWith(Sink.publisher())
+    val c = StreamTestKit.SubscriberProbe[Int]()
+    p.subscribe(c)
+    c.expectSubscriptionAndError().getMessage should be("no next")
+    c.expectNoMsg(100.millis)
+  }
+}
+
+class SynchronousIterableSpec extends AbstractFlowIteratorSpec {
```

Comment:
@drewhk here is the test for SynchronousIterablePublisher, I moved it here to be able to reuse things from AbstractFlowIteratorSpec